cf-check/backup.h

```diff
@@ -4,6 +4,6 @@
 #include <sequence.h>
 
 int backup_files(Seq *filenames);
-int backup_main(int argc, char **argv);
+int backup_main(int argc, const char *const *argv);
```

Comment:
@craigcomstock Generally, In C, the declaration(function signature usually in a .h file) and definition (implementation usually in a .c file) can have differences in parameter names and types, as long as the types are "compatible enough". This can be useful in some circumstances, and confusing in others.

**The declaration** gives you the information you need to know for calling the function. It gives the compiler adequate information to set up the arguments in registers / stack correctly without knowing how the function is implemented.

**The definition** is used for actually compiling the contents of the function, and enforcing the const restrictions withing the function scope, according to Cs rules.

So, consider this simple example:

```C
int double_it(int num);

int main(void)
{
  return double_it(3);
}

int double_it(const int n)
{
  // n = 2; // Nope, const..
  return 2 * n;
}
```

* The `double_it()` declaration here tells you that the function takes 1 integer (input) and returns 1 integer (output)
* You now know how to use the function within `main()`
* The `double_it()` definition has an added `const`, telling the compiler that the argument/local variable will not change within this function.

That's essentially it. The last asterisk in the pointer type refers to the local value, and I'm just declaring that it won't be assigned to in this function. It has 0 impact on the caller.

You can add it to the declaration, it makes no difference, but I prefer not to, because:

* If you just see the declaration, and you don't know about this stuff, it might be confusing why a function takes a const int (or const pointer).
* If somebody comes along later and removes it from the definition, it can still be in the declaration, without warnings. That makes it even more confusing.
* (I thought I've seen a warning for this once, essentially saying `this const does nothing`, but I can't remember where, maybe clang scan-build or something.)

The counterargument is of course that it can be confusing that the declaration and definition doesn't match, as evident by your question. So... tomato/potato.

**Note:** the param is called `n`/`num`. I don't recommend this, but I guess there are some _very rare_ situations where it's appropriate to have long, descriptive, names in the declaration and shorter more technical names in the definition. Maybe.

**Note 2:** Also, the function could have multiple definitions (inside `#ifdef`s). Some of them might want const, and other not. Different parameter names in the different definitions might also make sense.

Reply:
Perfect. Thank you for taking the time to explain so carefully. Very much appreciated. :1st_place_medal:

Reply:
Agreed. Few interesting points learned here!